{{cookiecutter.plugin_name}}/src/{{cookiecutter.module_name}}/_tests/test_widget.py

```diff
@@ -1,9 +1,54 @@
 import numpy as np
 
-from {{cookiecutter.module_name}} import ExampleQWidget, example_magic_widget
+from {{cookiecutter.module_name}}._widget import (
+    ExampleQWidget,
+    ImageThreshold,
+    example_function_widget,
+    example_magic_widget,
+)
+
+
+def test_example_function_widget():
+    # because our "widget" is a pure function, we can call it and
+    # test it independently of napari
+    im_data = np.random.random((100, 100))
+    thresholded = example_function_widget(im_data)
+    assert thresholded.shape == im_data.shape
+    # etc.
 
 
 # make_napari_viewer is a pytest fixture that returns a napari viewer object
+# you don't need to import it, as long as napari is installed
+# in your testing environment
+def test_example_magic_widget(make_napari_viewer):
+    viewer = make_napari_viewer()
+    layer = viewer.add_image(np.random.random((100, 100)))
+
+    # our widget will be a MagicFactory or FunctionGui instance
+    my_widget = example_magic_widget()
```

Comment:
Do we have to add the widget to the viewer with `viewer.add_dock_widget(...` or similar? 
Right now it's unclear what the purpose of having the viewer in the test is, since the actual assertion is just from calling the widget's function directly.

Reply:
Oh hang on, you pass `viewer.layers[0]` into the widget function. I see now. 
Ok, if this is the best test practice we want people to copy, I think this is fine. It certainly seems like an improvement over the last version, your extra comments here are great.